src/lib.rs

```diff
@@ -198,6 +204,66 @@ impl OpenOptions {
     }
 }
 
+/// Iterate over the contents of a directory. Created by calling read_dir() on
+/// an opened directory. Each item yielded by the iterator is an io::Result to
+/// allow communication of io errors as the iterator is advanced.
+///
+/// To the greatest extent possible the underlying OS semantics are preserved.
+/// That means that `.` and `..` entries are exposed, and that no sort order is
+/// guaranteed by the iterator.
+#[derive(Debug)]
+pub struct ReadDir<'a> {
+    _impl: ReadDirImpl<'a>,
+}
+
+impl<'a> ReadDir<'a> {
+    pub fn new(d: &'a mut File) -> Result<Self> {
+        Ok(ReadDir {
+            _impl: ReadDirImpl::new(d)?,
+        })
+    }
+}
+
+impl Iterator for ReadDir<'_> {
+    type Item = Result<DirEntry>;
+
+    fn next(&mut self) -> Option<Result<DirEntry>> {
+        self._impl
+            .next()
+            .map(|entry| entry.map(|_impl| DirEntry { _impl }))
+    }
+}
+
+/// The returned type for each entry found by [`read_dir`].
+///
+/// Each entry represents a single entry inside the directory. Platforms that
+/// provide rich metadata may in future expose this through methods or extension
+/// traits on DirEntry.
+///
+/// For now however, only the [`name()`] is exposed. This does not imply any
+/// additional IO for most workloads: metadata returned from a directory listing
+/// is inherently racy: presuming that what was a dir, or symlink etc when the
+/// directory was listed, will still be the same when opened is fallible.
+/// Instead, use open_at to open the contents, and then process based on the
+/// type of content found.
+#[derive(Debug)]
+pub struct DirEntry {
+    _impl: DirEntryImpl,
+}
+
+impl DirEntry {
+    pub fn name(&self) -> &OsStr {
+        self._impl.name()
+    }
+}
+
+/// Read the children of the directory d.
+///
+/// See [`ReadDir`] and [`DirEntry`] for details.
+pub fn read_dir(d: &mut File) -> Result<ReadDir> {
```

Comment:
I somewhat would have expected this would be a method on a specific `DirHandle` object, not a generic `fs::File`. Or, maybe it would be an extension trait of `File`... (I realize this is the style it's using now.)

Reply:
Yeah, so I can understand that.

There is a consistency question. The stdlib's OpenOptions struct is a programmable factory basically. I've chosen to put mkdir_at and open_at (and in future symlink_at and delete_at) on a similar struct, at least for now.  That permits a foreign File like that implements AsDeref<File> to be used I guess, which an extension trait wouldn't so much.

readdir doesn't seem to need OpenOptions which is why I made it a free function for now. 

The type it works on is File because one doesn't know when a path is opened whether it is a directory or not. We could do a type conversion thing where we check the metadata, but std lib File's don't have free metadata last I checked, so we'd be adding a syscall hidden behind a type conversion, which feels strange to me.

Reply:
You could have a `struct DirHandle(File)` that allows construction from a File without checking the on-disk type, just letting later calls fail if it's not really a dir.

It might make client code feel more natural: `dir_handle.readlink_at(name)`. Maybe this is imposing too much of an opinion on the client.

Of course on Linux (at least) it doesn't *have* to be a dir; you can have an `O_PATH`.

Reply:
perhaps. We can try a few things and see whats nice. e.g. 
```
struct FileAndError { f: File, e: io::Error}
DirHandle::from(File) -> Result<DirHandle, FileAndError>
```

would certainly work. I'm just not sure its better yet :)